docs/serialization.md

```diff
@@ -0,0 +1,98 @@
+- Root type discovery: logic to discover types which are considered for serialization
+- Type graph: recursive logic to build a set of types to consider for serialization, starting from the roots
+- Preservation logic: what the linker does with the discovered types
+
+## Root type discovery
+
+The heuristics will discover types that satisfy _all_ of the following criteria:
+
+- The type is used
+
+  There must be a statically discoverable reference to the type. In other words, if running the linker without the serialization heuristics removes a given type, then the heuristics will not keep it either.
+
+- The type or one of its members must be attributed with a serializer-specific attribute.
+
+  See the sections below about the attributes you can use for each serializer. The attribute must be present on the type, or one of its defined members, including fields/properties/methods/events, public or private, though the serializers may not define attributes that can be placed on all member kinds. It is not enough to place an attribute on a base type or on a member defined by a base type.
+
+Note that passing a type directly to a serializer constructor is _not_ enough to keep it. We do not use dataflow to discover types. For example:
+
+```csharp
+new XmlSerializer (typeof (RootType)); // Will not keep RootType
+```
+
+This pattern will not necessarily keep the type passed into the constructor, even though it is statically analyzable in theory.
+
+### XMLSerializer attributes
+
+On any member supported by the attribute:
+- Any attribute named `Xml*Attribute` in the `System.Xml.Serialization` namespace
+  - _except_ `XmlIgnoreAttribute`
+
+### DataContractSerializer attributes
+
+On types:
+- `System.Runtime.Serialization.DataContractAttribute`
+
+On properties, fields, or events:
+- `System.Runtime.Serialization.DataMemberAttribute`
+
+## Type graph
+
+Starting from the discovered root types, the heuristics will recursively discover the following types:
+
+- Base types
+  - including generic argument types
+- Types of public instance properties defined on the type
+  - a property is considered public if it has a public getter or setter
+  - including public properties of the base type
+  - including generic argument types
+- Types of public instance fields defined on the type
+  - including public fields of the base type
+  - including generic argument types
+
+Note that the types of implemented interfaces are not necessarily discovered.
+
+## Preservation logic
+
+For each discovered type (including root types and the recursive type graph), the linker marks the type and the following members:
+
+- Public instance properties
+  - including public or private getters and setters for such properties
+- Public instance fields
+- Public parameterless instance constructors
+
+Note that in general, private members and static members are not preserved, nor are methods or events (other than the mentioned constructor).
```

Comment:
What happens if I put `[DataMember]` on a private field - I would assume we preserve it... but this sentence seems to say otherwise.

Reply:
Currently it causes the type to be considered a root, but the field won't actually be kept. I think I can fix this without building serializer-specific logic into the recursive phase by marking such members up-front.

Reply:
Fixed in the latest iteration (see the commit comments and doc updates)